server/src/main/java/org/opensearch/gateway/BaseGatewayShardAllocator.java

```diff
@@ -74,7 +78,41 @@ public void allocateUnassigned(
         ExistingShardsAllocator.UnassignedAllocationHandler unassignedAllocationHandler
     ) {
         final AllocateUnassignedDecision allocateUnassignedDecision = makeAllocationDecision(shardRouting, allocation, logger);
+        executeDecision(shardRouting, allocateUnassignedDecision, allocation, unassignedAllocationHandler);
+    }
+
+    /**
+     * Allocate Set of unassigned shard  to nodes where valid copies of the shard already exists
+     * @param shards the shards to allocate
+     * @param allocation the allocation state container object
+     */
+    public void allocateUnassignedBatch(Set<ShardRouting> shards, RoutingAllocation allocation) {
+        // make Allocation Decisions for all shards
+        HashMap<ShardRouting, AllocateUnassignedDecision> decisionMap = makeAllocationDecision(shards, allocation, logger);
+        assert shards.size() == decisionMap.size() : "make allocation decision didn't return allocation decision for " + "some shards";
+        // get all unassigned shards iterator
+        RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
+
+        while (iterator.hasNext()) {
+            ShardRouting shard = iterator.next();
+            try {
+                if (decisionMap.isEmpty() == false) {
+                    if (decisionMap.containsKey(shard)) {
+                        executeDecision(shard, decisionMap.remove(shard), allocation, iterator);
+                    }
+                }
+            } catch (Exception e) {
+                logger.error("Failed to execute decision for shard {} ", shard, e);
+            }
+        }
```

Comment:
Set the `decisionMap` to `null` before exiting the method ?

Reply:
If `decisionMap.containsKey(shard)` condition is false, `decisionMap` will not be garbage collected.  

Would recommend reading about GC Reachability model for soft, weak and phantom references.

Reply:
@amkhar Answered the role of DecisionMap [here](https://github.com/opensearch-project/OpenSearch/pull/8776/files#r1413458388)

Once the method run[allocateUnassignedBatch] is over and exited, decisionMap being a local variable will be garbage collected?

Reply:
Yes. It'll be garbage collected.